staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types_discovery.go

```diff
@@ -0,0 +1,144 @@
+	// from the kubectl CLI interface.
+	SingularName string `json:"singularName" protobuf:"bytes,4,opt,name=singularName"`
+	// Verbs is a list of supported kube verbs (this includes get, list, watch, create,
+	// update, patch, delete, deletecollection, other)
+	Verbs Verbs `json:"verbs" protobuf:"bytes,5,opt,name=verbs"`
+	// ShortNames is a list of suggested short names of the resource.
+	// +listType=set
+	ShortNames []string `json:"shortNames,omitempty" protobuf:"bytes,6,rep,name=shortNames"`
+	// Categories is a list of the grouped resources this resource belongs to (e.g. 'all')
+	// +listType=set
+	Categories []string `json:"categories,omitempty" protobuf:"bytes,7,rep,name=categories"`
+	// Subresources is a list of subresources provided by this resource.  Subresources are located at
+	// +listType=map
+	// +listMapKey=subresource
+	Subresources []APISubresourceDiscovery `json:"subresources,omitempty" protobuf:"bytes,8,rep,name=subresources"`
+}
+
+// ResourceScope is an enum defining the different scopes available to a resource.
+type ResourceScope string
+
+const (
+	ScopeCluster   ResourceScope = "Cluster"
+	ScopeNamespace ResourceScope = "Namespace"
+)
+
+// APIDiscoveryKind provides the group version kind information for a resource.
+type APIDiscoveryKind struct {
+	// group is the group of the serialization.
+	Group string `json:"group,omitempty" protobuf:"bytes,1,opt,name=group"`
+	// version is the version of the serialization
+	Version string `json:"version,omitempty" protobuf:"bytes,2,opt,name=version"`
+	// kind is the kind of the serialiation
+	Kind string `json:"kind" protobuf:"bytes,3,opt,name=kind"`
+}
+
+// APISubresourceDiscovery provides information about an API subresource for discovery.
+type APISubresourceDiscovery struct {
+	// subresource is the name of the subresource.  This is used in the URL path and is the unique identifier
+	// for this resource across all versions.
+	Subresource string `json:"subresource" protobuf:"bytes,1,opt,name=subresource"`
+	// returnType describes the type of serialization that will be returned from this endpoint.
+	// Some subresources do not return normal resources, these will have nil return types.
+	ReturnType APIDiscoveryKind `json:"returnType,omitempty" protobuf:"bytes,2,opt,name=returnType"`
+	// acceptedTypes describes the kinds that this endpoint accepts.  It is possible for a subresource to accept multiple kinds.
+	// It is also possible for an endpoint to accept no standard types.  Those will have a zero length list.
+	// +listType=set
+	AcceptedTypes []APIDiscoveryKind `json:"acceptedTypes,omitempty" protobuf:"bytes,3,rep,name=acceptedTypes"`
+	// verbs is a list of supported kube verbs (this includes get, list, watch, create,
+	// update, patch, delete, other)
+	Verbs Verbs `json:"verbs" protobuf:"bytes,4,opt,name=verbs"`
+}
+
+// APIVersionDiscoveryStatus contains the observations of an APIGroupDisocvery
+type APIVersionDiscoveryStatus struct {
```

Comment:
if we added a `StaleVersions []string` to this type, would that allow us to keep Status on APIGroupDiscovery instead of repeated on subelements?